process/interceptors/factory/eligibleNodesCache.go

```diff
@@ -0,0 +1,243 @@
+	mutEligibleNodesMap     sync.RWMutex
+
+	peerShardMapper  process.PeerShardMapper
+	nodesCoordinator process.NodesCoordinator
+}
+
+func newEligibleNodesCache(
+	peerShardMapper process.PeerShardMapper,
+	nodesCoordinator process.NodesCoordinator,
+) (*eligibleNodesCache, error) {
+	if check.IfNil(peerShardMapper) {
+		return nil, process.ErrNilPeerShardMapper
+	}
+	if check.IfNil(nodesCoordinator) {
+		return nil, process.ErrNilNodesCoordinator
+	}
+
+	return &eligibleNodesCache{
+		eligibleNodesMap:        map[uint32]map[uint32]map[string]struct{}{},
+		epochsLimitsForShardMap: map[uint32]*epochLimits{},
+		peerShardMapper:         peerShardMapper,
+		nodesCoordinator:        nodesCoordinator,
+	}, nil
+}
+
+// IsPeerEligible returns true if the provided peer is eligible
+func (cache *eligibleNodesCache) IsPeerEligible(pid core.PeerID, shard uint32, epoch uint32) bool {
+	isPeerEligible, shouldUpdateCache := cache.isPeerEligible(pid, shard, epoch)
+	if !shouldUpdateCache {
+		return isPeerEligible
+	}
+
+	cache.mutEligibleNodesMap.Lock()
+	defer cache.mutEligibleNodesMap.Unlock()
+
+	// get the eligible list for the new epoch from nodesCoordinator
+	eligibleNodesForShardInEpoch, err := cache.getEligibleNodesForShardInEpoch(epoch, shard)
+	if err != nil {
+		log.Trace("IsPeerEligible.getEligibleNodesForShardInEpoch failed", "error", err)
+		return false
+	}
+
+	cache.updateMapsIfNeeded(shard, epoch, eligibleNodesForShardInEpoch)
+
+	// check the eligible list from the new proof's epoch and shard
+	peerInfo := cache.peerShardMapper.GetPeerInfo(pid)
+	_, isNodeEligible := eligibleNodesForShardInEpoch[string(peerInfo.PkBytes)]
+
+	return isNodeEligible
+}
+
+func (cache *eligibleNodesCache) updateMapsIfNeeded(
+	shard uint32,
+	newEpoch uint32,
+	eligibleNodesForShardInEpoch map[string]struct{},
+) {
+	// if the shard is new, create the map and store the received epoch
+	limitEpochsForShard, isShardCached := cache.epochsLimitsForShardMap[shard]
+	if !isShardCached {
+		cache.epochsLimitsForShardMap[shard] = &epochLimits{
+			min: newEpoch,
+			max: newEpoch,
+		}
+		cache.eligibleNodesMap[shard] = map[uint32]map[string]struct{}{
+			newEpoch: eligibleNodesForShardInEpoch,
+		}
+
+		return
+	}
+
+	// reaching this point means that the shard is cached, but we have a new epoch
+	// there are different situations for the new epoch:
+	//	1. the epoch is higher than the min cached epoch + epochs delta ->
+	//		store the new + all cached ones that ate still in delta limit
+	//  2. the epoch is lower than the max cached epoch - epochs delta ->
+	//		store the new + all cached ones that ate still in delta limit
+	// 	3. the new epoch is higher than the min cached but still in the delta limits ->
+	//		cache it and update the max cached if needed
+	//	4. the new epoch is lower than the max cached but still in the delta limits ->
+	//		cache it and update the min cached if needed
+
+	// case 1.
+	if limitEpochsForShard.isHigherOutOfRange(newEpoch) {
```

Comment:
separate functions for these cases? at least for the first 2

Reply:
extracted all cases into separate methods